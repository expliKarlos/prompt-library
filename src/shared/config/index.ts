import * as fs from 'fs';

import { CommonConfig, commonConfig } from './common-config';
import { CONFIG_DIR, CONFIG_FILE, isCliEnvironment } from './constants';
import { appConfig, AppConfig } from '../../app/config/app-config';
import { CliConfig, cliConfig } from '../../cli/config/cli-config';

export type Config = CommonConfig & (CliConfig | AppConfig);

let loadedConfig: Config | null = null;
let lastCliEnv: string | undefined;

export function clearConfigCache(): void {
    loadedConfig = null;
    lastCliEnv = undefined;
}

function loadConfig(): Config {
    if (process.env.CLI_ENV !== lastCliEnv) {
        loadedConfig = null;
        lastCliEnv = process.env.CLI_ENV;
    }

    if (loadedConfig) {
        return JSON.parse(JSON.stringify(loadedConfig));
    }

    const environmentConfig = isCliEnvironment ? cliConfig : appConfig;
    let config = {
        ...commonConfig,
        ...environmentConfig
    } as Config;

    if (isCliEnvironment && fs.existsSync(CONFIG_FILE)) {
        try {
            const fileConfig = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
            config = { ...config, ...fileConfig };
        } catch (error) {
            console.error('Error loading config file:', error);
        }
    }

    loadedConfig = JSON.parse(JSON.stringify(config));
    return config;
}

export function getConfig(): Readonly<Config> {
    return loadConfig();
}

export function setConfig<K extends keyof Config>(key: K, value: Config[K]): void {
    if (process.env.CLI_ENV !== 'cli') {
        throw new Error('setConfig is only available in CLI environment');
    }

    if (!fs.existsSync(CONFIG_DIR)) {
        fs.mkdirSync(CONFIG_DIR, { recursive: true });
    }

    const config = loadConfig();
    const updatedConfig = { ...config, [key]: value };
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(updatedConfig, null, 2));

    loadedConfig = updatedConfig;

    if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
        process.env[key] = String(value);
    }
}

export function getConfigValue<K extends keyof Config>(key: K): Config[K] {
    const config = loadConfig();

    if (!isCliEnvironment && process.env[key] !== undefined) {
        const envValue = process.env[key];
        const currentValue = config[key];

        if (typeof currentValue === 'number') {
            return Number(envValue) as Config[K];
        } else if (typeof currentValue === 'boolean') {
            return (envValue?.toLowerCase() === 'true') as unknown as Config[K];
        }
        return envValue as Config[K];
    }
    return config[key];
}
