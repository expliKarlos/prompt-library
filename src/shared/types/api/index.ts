/**
 * Re-export all API-related types
 */
export * from './result';
