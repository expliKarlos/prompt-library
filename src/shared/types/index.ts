export interface EnvVariable {
    id: number;
    name: string;
    value: string;
    scope: 'global' | 'prompt';
    prompt_id?: number;
}

export interface CategoryItem {
    id: string;
    title: string;
    primary_category: string;
    description: string;
    path: string;
    subcategories: string[];
}

export interface PromptVariable {
    name: string;
    role: string;
    optional_for_user: boolean;
    value?: string;
}

export type ApiResult<T> = {
    success: boolean;
    data?: T;
    error?: string;
};

export interface PromptMetadata {
    id?: string;
    title: string;
    primary_category: string;
    subcategories: string[];
    directory: string;
    tags: string | string[];
    one_line_description: string;
    description: string;
    variables: PromptVariable[];
    content_hash?: string;
    fragments?: PromptFragment[];
}

export interface PromptFragment {
    name: string;
    category: string;
    variable?: string;
}
