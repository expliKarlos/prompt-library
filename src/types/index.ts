/* eslint-disable */
export * from './fragment';
export * from './metadata';
