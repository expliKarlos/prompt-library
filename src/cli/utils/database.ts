import path from 'path';

import fs from 'fs-extra';
import yaml from 'js-yaml';
import NodeCache from 'node-cache';
import sqlite3, { RunResult } from 'sqlite3';

import { AppError, handleError } from './errors';
import { createPrompt } from './prompts';
import { commonConfig } from '../../shared/config/common-config';
import { ApiResult, CategoryItem, PromptMetadata, PromptVariable } from '../../shared/types';
import { fileExists, readDirectory, readFileContent } from '../../shared/utils/file-system';
import logger from '../../shared/utils/logger';
import { cliConfig } from '../config/cli-config';

const db = new sqlite3.Database(cliConfig.DB_PATH);

export const cache = new NodeCache({ stdTTL: 600 });

export async function runAsync(sql: string, params: any[] = []): Promise<ApiResult<RunResult>> {
    return new Promise((resolve) => {
        db.run(sql, params, function (this: RunResult, err) {
            if (err) {
                handleError(new AppError('DB_ERROR', err.message), 'runAsync');
                resolve({ success: false, error: err.message });
            } else {
                resolve({ success: true, data: this });
            }
        });
    });
}

export function getAsync<T = any>(sql: string, params: any[] = []): Promise<ApiResult<T>> {
    return new Promise((resolve) => {
        db.get(sql, params, (err, row) => {
            if (err) {
                handleError(new AppError('DB_ERROR', err.message), 'getAsync');
                resolve({ success: false, error: err.message });
            } else {
                resolve({ success: true, data: row as T });
            }
        });
    });
}

export function allAsync<T = any>(sql: string, params: any[] = []): Promise<ApiResult<T[]>> {
    return new Promise((resolve) => {
        db.all(sql, params, (err, rows) => {
            if (err) {
                handleError(new AppError('DB_ERROR', err.message), 'allAsync');
                resolve({ success: false, error: err.message });
            } else {
                resolve({ success: true, data: rows as T[] });
            }
        });
    });
}

export async function handleApiResult<T>(result: ApiResult<T>, message: string): Promise<T | null> {
    if (result.success && result.data) {
        return result.data;
    } else {
        handleError(result.error || 'Unknown error', `API call in ${message}`);
        return null;
    }
}

export async function getCachedOrFetch<T>(
    key: string,
    fetchFn: () => Promise<ApiResult<T>>,
    cacheInstance: NodeCache = cache
): Promise<ApiResult<T>> {
    const cachedResult = cacheInstance.get<T>(key);

    if (cachedResult !== undefined) {
        return { success: true, data: cachedResult };
    }

    const result = await fetchFn();

    if (result.success && result.data) {
        cacheInstance.set(key, result.data);
    }
    return result;
}

export async function initDatabase(): Promise<ApiResult<void>> {
    try {
        await fs.ensureDir(path.dirname(cliConfig.DB_PATH));

        const tables = [
            `CREATE TABLE IF NOT EXISTS prompts (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                content TEXT NOT NULL,
                primary_category TEXT NOT NULL,
                directory TEXT NOT NULL,
                one_line_description TEXT,
                description TEXT,
                content_hash TEXT,
                tags TEXT,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
            )`,
            `CREATE TABLE IF NOT EXISTS subcategories (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                prompt_id INTEGER,
                name TEXT NOT NULL,
                FOREIGN KEY (prompt_id) REFERENCES prompts (id)
            )`,
            `CREATE TABLE IF NOT EXISTS variables (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                prompt_id INTEGER,
                name TEXT NOT NULL,
                role TEXT NOT NULL,
                optional_for_user BOOLEAN NOT NULL,
                value TEXT DEFAULT '',
                FOREIGN KEY (prompt_id) REFERENCES prompts (id)
            )`,
            `CREATE TABLE IF NOT EXISTS fragments (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                prompt_id INTEGER,
                category TEXT NOT NULL,
                name TEXT NOT NULL,
                variable TEXT NOT NULL DEFAULT '',
                FOREIGN KEY (prompt_id) REFERENCES prompts (id)
            )`,
            `CREATE TABLE IF NOT EXISTS env_vars (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL UNIQUE,
                value TEXT NOT NULL,
                scope TEXT NOT NULL,
                prompt_id INTEGER,
                FOREIGN KEY (prompt_id) REFERENCES prompts (id)
            )`,
            `CREATE TABLE IF NOT EXISTS prompt_executions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                prompt_id INTEGER,
                execution_time DATETIME DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (prompt_id) REFERENCES prompts (id)
            )`,
            `CREATE TABLE IF NOT EXISTS favorite_prompts (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                prompt_id INTEGER,
                added_time DATETIME DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (prompt_id) REFERENCES prompts (id),
                UNIQUE(prompt_id)
            )`
        ];

        for (const table of tables) {
            const result = await runAsync(table);

            if (!result.success) {
                return { success: false, error: result.error };
            }
        }
        return { success: true };
    } catch (error) {
        handleError(error, 'initializing database');
        return { success: false, error: 'Failed to initialize database' };
    }
}

export async function fetchCategories(): Promise<ApiResult<Record<string, CategoryItem[]>>> {
    return getCachedOrFetch('all_prompts_categories', async () => {
        const result = await allAsync<any>(`
            SELECT p.id, p.title, p.primary_category, p.description, p.directory as path, p.tags,
                   GROUP_CONCAT(DISTINCT s.name) as subcategories
            FROM prompts p
            LEFT JOIN subcategories s ON p.id = s.prompt_id
            GROUP BY p.id
        `);

        if (!result.success) {
            return { success: false, error: 'Failed to fetch prompts with categories' };
        }

        const categorizedPrompts: Record<string, CategoryItem[]> = {};
        result.data?.forEach((prompt) => {
            const category = prompt.primary_category;

            if (!categorizedPrompts[category]) {
                categorizedPrompts[category] = [];
            }

            categorizedPrompts[category].push({
                ...prompt,
                tags: prompt.tags ? prompt.tags.split(',') : [],
                subcategories: prompt.subcategories ? prompt.subcategories.split(',') : []
            });
        });
        return { success: true, data: categorizedPrompts };
    });
}

export async function getPromptDetails(
    promptId: string
): Promise<ApiResult<PromptMetadata & { variables: PromptVariable[] }>> {
    const promptResult = await getAsync<PromptMetadata>('SELECT * FROM prompts WHERE id = ?', [promptId]);
    const variablesResult = await allAsync<PromptVariable>(
        'SELECT name, role, value, optional_for_user FROM variables WHERE prompt_id = ?',
        [promptId]
    );

    if (!promptResult.success || !variablesResult.success) {
        return { success: false, error: 'Failed to fetch prompt details' };
    }

    if (promptResult.data) {
        promptResult.data.tags = promptResult.data.tags || [];
    }
    return {
        success: true,
        data: {
            ...promptResult.data!,
            variables: variablesResult.data || []
        }
    };
}

export async function updatePromptVariable(
    promptId: string,
    variableName: string,
    value: string
): Promise<ApiResult<void>> {
    try {
        const result = await runAsync('UPDATE variables SET value = ? WHERE prompt_id = ? AND name = ?', [
            value,
            promptId,
            variableName
        ]);

        if (!result.success) {
            return { success: false, error: result.error || 'Unknown database error' };
        }

        if (result.data?.changes === 0) {
            return { success: false, error: `No variable found with name ${variableName} for prompt ${promptId}` };
        }

        cache.flushAll();
        return { success: true };
    } catch (error) {
        handleError(error, 'updating prompt variable');
        return { success: false, error: 'Failed to update prompt variable' };
    }
}

export async function getRecentExecutions(limit: number = 5): Promise<any[]> {
    const query = `
        SELECT pe.id, pe.prompt_id, pe.execution_time, p.title, p.primary_category 
        FROM prompt_executions pe
        JOIN prompts p ON pe.prompt_id = p.id
        ORDER BY pe.execution_time DESC
        LIMIT ?
    `;

    try {
        const result = await allAsync(query, [limit]);

        if (result.success && result.data && result.data.length > 0) {
            return result.data;
        }
        return [];
    } catch (error) {
        handleError(error, 'fetching recent executions');
        return [];
    }
}

export async function recordPromptExecution(promptId: number | string): Promise<ApiResult<void>> {
    try {
        const result = await runAsync('INSERT INTO prompt_executions (prompt_id) VALUES (?)', [promptId]);
        return { success: result.success };
    } catch (error) {
        handleError(error, 'recording prompt execution');
        return { success: false, error: 'Failed to record prompt execution' };
    }
}

export async function addPromptToFavorites(promptId: number | string): Promise<ApiResult<void>> {
    try {
        const result = await runAsync('INSERT OR IGNORE INTO favorite_prompts (prompt_id) VALUES (?)', [promptId]);
        return { success: result.success };
    } catch (error) {
        handleError(error, 'adding prompt to favorites');
        return { success: false, error: 'Failed to add prompt to favorites' };
    }
}

export async function removePromptFromFavorites(promptId: number | string): Promise<ApiResult<void>> {
    try {
        const result = await runAsync('DELETE FROM favorite_prompts WHERE prompt_id = ?', [promptId]);
        return { success: result.success };
    } catch (error) {
        handleError(error, 'removing prompt from favorites');
        return { success: false, error: 'Failed to remove prompt from favorites' };
    }
}

export async function getFavoritePrompts(): Promise<ApiResult<any[]>> {
    const query = `
        SELECT fp.id, fp.prompt_id, fp.added_time, p.title, p.primary_category, p.description
        FROM favorite_prompts fp
        JOIN prompts p ON fp.prompt_id = p.id
        ORDER BY fp.added_time DESC
    `;

    try {
        return await allAsync(query);
    } catch (error) {
        handleError(error, 'fetching favorite prompts');
        return { success: false, error: 'Failed to fetch favorite prompts' };
    }
}

export async function syncPromptsWithDatabase(): Promise<ApiResult<void>> {
    try {
        const promptDirs = await readDirectory(cliConfig.PROMPTS_DIR);
        logger.info(`Found ${promptDirs.length} items in prompts directory:`, promptDirs);

        await runAsync('DELETE FROM prompts', []);
        await runAsync('DELETE FROM subcategories', []);
        await runAsync('DELETE FROM variables', []);
        await runAsync('DELETE FROM fragments', []);

        for (const dir of promptDirs) {
            const promptPath = path.join(cliConfig.PROMPTS_DIR, dir, commonConfig.PROMPT_FILE_NAME);
            const metadataPath = path.join(cliConfig.PROMPTS_DIR, dir, commonConfig.METADATA_FILE_NAME);
            logger.info(`Checking prompt: ${dir}`);
            logger.info(`Prompt file exists: ${await fileExists(promptPath)}`);
            logger.info(`Metadata file exists: ${await fileExists(metadataPath)}`);

            if ((await fileExists(promptPath)) && (await fileExists(metadataPath))) {
                const promptContent = await readFileContent(promptPath);
                const metadataContent = await readFileContent(metadataPath);

                try {
                    const metadata = yaml.load(metadataContent) as PromptMetadata;
                    await createPrompt(metadata, promptContent);
                    logger.info(`Successfully processed prompt: ${dir}`);
                } catch (error) {
                    handleError(error, `processing prompt ${dir}`);
                }
            } else {
                logger.warn(`Skipping ${dir}: missing prompt or metadata file`);
            }
        }

        cache.flushAll();

        logger.info('Database sync completed');
        return { success: true };
    } catch (error) {
        handleError(error, 'syncing prompts with database');
        return { success: false, error: 'Failed to sync prompts with database' };
    }
}

export async function cleanupOrphanedData(): Promise<ApiResult<void>> {
    try {
        const promptDirs = await readDirectory(cliConfig.PROMPTS_DIR);
        const existingPromptIds = promptDirs.map((dir) => parseInt(dir, 10)).filter((id) => !isNaN(id));

        if (existingPromptIds.length === 0) {
            logger.warn('No valid prompt IDs found. Skipping orphaned data cleanup.');
            return { success: true };
        }

        await runAsync('DELETE FROM prompts WHERE id NOT IN (?)', [existingPromptIds.join(',')]);
        await runAsync('DELETE FROM subcategories WHERE prompt_id NOT IN (SELECT id FROM prompts)');

        const orphanedVariables = await allAsync<{ id: number; name: string; prompt_id: number }>(
            'SELECT id, name, prompt_id FROM variables WHERE prompt_id NOT IN (SELECT id FROM prompts)'
        );

        if (orphanedVariables.success && orphanedVariables.data) {
            for (const variable of orphanedVariables.data) {
                const usedVariable = await getAsync<{ count: number }>(
                    'SELECT COUNT(*) as count FROM variables WHERE name = ? AND prompt_id IN (SELECT id FROM prompts)',
                    [variable.name]
                );

                if (usedVariable.success && usedVariable.data && usedVariable.data.count === 0) {
                    await runAsync('DELETE FROM variables WHERE id = ?', [variable.id]);
                    logger.info(`Deleted orphaned variable: ${variable.name}`);
                } else {
                    logger.info(`Kept shared variable: ${variable.name}`);
                }
            }
        }

        await runAsync('DELETE FROM fragments WHERE prompt_id NOT IN (SELECT id FROM prompts)');

        const orphanedEnvVars = await allAsync<{ id: number; name: string; prompt_id: number | null }>(
            'SELECT id, name, prompt_id FROM env_vars WHERE prompt_id IS NOT NULL AND prompt_id NOT IN (SELECT id FROM prompts)'
        );

        if (orphanedEnvVars.success && orphanedEnvVars.data) {
            for (const envVar of orphanedEnvVars.data) {
                const usedEnvVar = await getAsync<{ count: number }>(
                    'SELECT COUNT(*) as count FROM env_vars WHERE name = ? AND (prompt_id IS NULL OR prompt_id IN (SELECT id FROM prompts))',
                    [envVar.name]
                );

                if (usedEnvVar.success && usedEnvVar.data && usedEnvVar.data.count === 0) {
                    await runAsync('DELETE FROM env_vars WHERE id = ?', [envVar.id]);
                    logger.info(`Deleted orphaned env var: ${envVar.name}`);
                } else {
                    logger.info(`Kept shared env var: ${envVar.name}`);
                }
            }
        }

        cache.flushAll();

        logger.info('Orphaned data cleaned up successfully');
        return { success: true };
    } catch (error) {
        handleError(error, 'cleaning up orphaned data');
        return { success: false, error: 'Failed to clean up orphaned data' };
    }
}

export async function flushData(): Promise<void> {
    logger.info('Starting data flush process');

    try {
        await runAsync('DELETE FROM prompts');
        await runAsync('DELETE FROM subcategories');
        await runAsync('DELETE FROM variables');
        await runAsync('DELETE FROM fragments');
        await runAsync('DELETE FROM env_vars');
        logger.info('Database tables cleared');

        await fs.emptyDir(cliConfig.PROMPTS_DIR);
        logger.info('Prompts directory cleared');

        await fs.emptyDir(cliConfig.FRAGMENTS_DIR);
        logger.info('Fragments directory cleared');

        await fs.emptyDir(cliConfig.TEMP_DIR);
        logger.info('Temporary directory cleared');

        logger.info('Data flush completed successfully');
    } catch (error) {
        handleError(error, 'flushing data');
        throw new Error('Failed to flush data');
    }
}

export { db };
