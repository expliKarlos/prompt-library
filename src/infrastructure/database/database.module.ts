import { Module, forwardRef } from '@nestjs/common';

import { DatabaseRepository } from './repositories/database.repository';
import { SqliteFragmentRepository } from './repositories/sqlite-fragment.repository';
import { SqlitePromptExecutionRepository } from './repositories/sqlite-prompt-execution.repository';
import { SqlitePromptFavoriteRepository } from './repositories/sqlite-prompt-favorite.repository';
import { SqlitePromptMetadataRepository } from './repositories/sqlite-prompt-metadata.repository';
import { SqlitePromptSyncRepository } from './repositories/sqlite-prompt-sync.repository';
import { SqlitePromptRepository } from './repositories/sqlite-prompt.repository';
import { SqliteVariableRepository } from './repositories/sqlite-variable.repository';
import { DatabaseService } from './services/database.service';
import { FlushCommandService } from './services/flush.service';
import { IFragmentRepository } from '../../core/fragment/repositories/fragment.repository.interface';
import { IPromptExecutionRepository } from '../../core/prompt/repositories/prompt-execution.repository.interface';
import { IPromptFavoriteRepository } from '../../core/prompt/repositories/prompt-favorite.repository.interface';
import { IPromptMetadataRepository } from '../../core/prompt/repositories/prompt-metadata.repository.interface';
import { IPromptSyncRepository } from '../../core/prompt/repositories/prompt-sync.repository.interface';
import { IPromptRepository } from '../../core/prompt/repositories/prompt.repository.interface';
import { IVariableRepository } from '../../core/variable/repositories/variable.repository.interface';
import { CommonInfrastructureModule } from '../common/common-infrastructure.module';
import { ErrorModule } from '../error/error.module';
import { LoggerModule } from '../logger/logger.module';

@Module({
    imports: [
        forwardRef(() => LoggerModule),
        forwardRef(() => ErrorModule),
        forwardRef(() => CommonInfrastructureModule)
    ],
    providers: [
        DatabaseRepository,
        DatabaseService,
        FlushCommandService,
        SqliteFragmentRepository,
        SqlitePromptRepository,
        SqlitePromptMetadataRepository,
        SqlitePromptExecutionRepository,
        SqlitePromptFavoriteRepository,
        SqlitePromptSyncRepository,
        SqliteVariableRepository,
        { provide: IFragmentRepository, useExisting: SqliteFragmentRepository },
        { provide: IPromptRepository, useExisting: SqlitePromptRepository },
        { provide: IPromptMetadataRepository, useExisting: SqlitePromptMetadataRepository },
        { provide: IPromptExecutionRepository, useExisting: SqlitePromptExecutionRepository },
        { provide: IPromptFavoriteRepository, useExisting: SqlitePromptFavoriteRepository },
        { provide: IPromptSyncRepository, useExisting: SqlitePromptSyncRepository },
        { provide: IVariableRepository, useExisting: SqliteVariableRepository }
    ],
    exports: [
        DatabaseService,
        DatabaseRepository,
        FlushCommandService,
        IFragmentRepository,
        IPromptRepository,
        IPromptMetadataRepository,
        IPromptExecutionRepository,
        IPromptFavoriteRepository,
        IPromptSyncRepository,
        IVariableRepository
    ]
})
export class DatabaseModule {}
