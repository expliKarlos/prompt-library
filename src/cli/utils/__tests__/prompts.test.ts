/**
 * Improved test suite for prompts utility
 * This combines the best of both approaches between prompts.test.ts and prompts-simple.test.ts
 */
import { RunResult } from 'sqlite3';

import { PromptMetadata } from '../../../shared/types';
import { allAsync, getAsync, runAsync } from '../database';
import { readEnvVars } from '../env-vars';
import { createPrompt, listPrompts, getPromptFiles, getPromptMetadata, viewPromptDetails } from '../prompts';

jest.mock('../database');
jest.mock('../env-vars');
jest.mock('chalk', () => ({
    cyan: jest.fn((text) => text),
    green: jest.fn((text) => text),
    blue: jest.fn((text) => text),
    magenta: jest.fn((text) => text),
    yellow: jest.fn((text) => text),
    red: jest.fn((text) => text)
}));
jest.mock('../errors', () => ({
    handleError: jest.fn()
}));

jest.mock('../database');
const mockGetAsync = getAsync as jest.MockedFunction<typeof getAsync>;
const mockAllAsync = allAsync as jest.MockedFunction<typeof allAsync>;
const mockRunAsync = runAsync as jest.MockedFunction<typeof runAsync>;
const mockMetadata: PromptMetadata = {
    title: 'Test Prompt',
    primary_category: 'test',
    subcategories: ['sub1', 'sub2'],
    directory: 'test-dir',
    tags: ['tag1', 'tag2'],
    one_line_description: 'Test description',
    description: 'Full test description',
    variables: [
        { name: 'var1', role: 'test role', optional_for_user: false },
        { name: 'var2', role: 'test role 2', optional_for_user: true }
    ],
    content_hash: 'test-hash',
    fragments: [{ category: 'test', name: 'fragment1', variable: '{{var1}}' }]
};
describe('PromptsUtils', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.spyOn(console, 'log').mockImplementation(() => {});
        jest.spyOn(console, 'error').mockImplementation(() => {});
    });

    describe('createPrompt', () => {
        it('should successfully create a prompt with all metadata', async () => {
            mockRunAsync.mockResolvedValueOnce({
                success: true,
                data: { lastID: 1, changes: 1 } as unknown as RunResult
            });

            mockRunAsync.mockResolvedValue({
                success: true,
                data: { lastID: 1, changes: 1 } as unknown as RunResult
            });

            const result = await createPrompt(mockMetadata, 'Test content');
            expect(result.success).toBe(true);
            expect(mockRunAsync).toHaveBeenCalledWith(
                expect.stringContaining('INSERT INTO prompts'),
                expect.arrayContaining([mockMetadata.title])
            );
        });

        it('should handle database errors gracefully', async () => {
            mockRunAsync.mockResolvedValueOnce({
                success: false,
                error: 'DB error'
            });

            const result = await createPrompt(mockMetadata, 'Test content');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Failed to insert prompt');
        });
    });

    describe('listPrompts', () => {
        it('should return list of prompts successfully', async () => {
            const mockPrompts = [
                { id: 1, title: 'Prompt 1', primary_category: 'cat1' },
                { id: 2, title: 'Prompt 2', primary_category: 'cat2' }
            ];
            mockAllAsync.mockResolvedValueOnce({
                success: true,
                data: mockPrompts
            });

            const result = await listPrompts();
            expect(result.success).toBe(true);
            expect(result.data).toEqual(mockPrompts);
        });

        it('should handle database errors', async () => {
            mockAllAsync.mockResolvedValueOnce({
                success: false,
                error: 'DB error'
            });

            const result = await listPrompts();
            expect(result.success).toBe(false);
            expect(result.error).toBe('Failed to list prompts');
        });
    });

    describe('getPromptFiles', () => {
        it('should return prompt content and metadata', async () => {
            mockGetAsync.mockImplementation((query) => {
                if (query.includes('SELECT id, content FROM prompts')) {
                    return Promise.resolve({
                        success: true,
                        data: { id: '1', content: 'Test content' }
                    });
                } else if (query.includes('SELECT * FROM prompts')) {
                    return Promise.resolve({
                        success: true,
                        data: {
                            id: '1',
                            title: 'Test Prompt',
                            primary_category: 'test',
                            directory: 'test-dir',
                            one_line_description: 'Test description',
                            description: 'Full test description',
                            content_hash: 'test-hash',
                            tags: 'tag1,tag2'
                        }
                    });
                }
                return Promise.resolve({ success: false });
            });

            mockAllAsync.mockImplementation((query) => {
                if (query.includes('subcategories')) {
                    return Promise.resolve({
                        success: true,
                        data: [{ name: 'sub1' }, { name: 'sub2' }]
                    });
                } else if (query.includes('variables')) {
                    return Promise.resolve({
                        success: true,
                        data: [
                            { name: 'var1', role: 'test role', optional_for_user: false },
                            { name: 'var2', role: 'test role 2', optional_for_user: true }
                        ]
                    });
                } else if (query.includes('fragments')) {
                    return Promise.resolve({
                        success: true,
                        data: [{ category: 'test', name: 'fragment1', variable: '{{var1}}' }]
                    });
                }
                return Promise.resolve({ success: false });
            });

            const result = await getPromptFiles('1');
            expect(result.success).toBe(true);
            expect(result.data?.promptContent).toBe('Test content');
            expect(result.data?.metadata).toBeDefined();
        });

        it('should return error when prompt is not found', async () => {
            mockGetAsync.mockResolvedValueOnce({ success: false }).mockResolvedValueOnce({ success: false });

            const result = await getPromptFiles('missing');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Prompt not found');
        });
    });

    describe('getPromptMetadata', () => {
        it('should return complete metadata when all data is available', async () => {
            mockGetAsync.mockResolvedValueOnce({
                success: true,
                data: {
                    id: '1',
                    title: 'Test Prompt',
                    primary_category: 'test',
                    directory: 'test-dir',
                    one_line_description: 'Test description',
                    description: 'Full test description',
                    content_hash: 'test-hash',
                    tags: 'tag1,tag2'
                }
            });

            mockAllAsync
                .mockResolvedValueOnce({
                    success: true,
                    data: [{ name: 'sub1' }, { name: 'sub2' }]
                })
                .mockResolvedValueOnce({
                    success: true,
                    data: [
                        { name: 'var1', role: 'test role', optional_for_user: false },
                        { name: 'var2', role: 'test role 2', optional_for_user: true }
                    ]
                })
                .mockResolvedValueOnce({
                    success: true,
                    data: [{ category: 'test', name: 'fragment1', variable: '{{var1}}' }]
                });

            const result = await getPromptMetadata('1');
            expect(result.success).toBe(true);
            expect(result.data?.title).toBe('Test Prompt');
            expect(result.data?.subcategories).toHaveLength(2);
            expect(result.data?.variables).toHaveLength(2);
            expect(result.data?.fragments).toHaveLength(1);
        });

        it('should return error when metadata is not found', async () => {
            mockGetAsync.mockResolvedValueOnce({
                success: false,
                error: 'Not found'
            });

            const result = await getPromptMetadata('999');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Metadata not found');
        });

        it('should return error when subcategories retrieval fails', async () => {
            mockGetAsync.mockResolvedValueOnce({
                success: true,
                data: { id: '1', title: 'Test' }
            });

            mockAllAsync.mockResolvedValueOnce({
                success: false,
                error: 'Failed to get subcategories'
            });

            const result = await getPromptMetadata('1');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Failed to get subcategories');
        });

        it('should return error when variables retrieval fails', async () => {
            mockGetAsync.mockResolvedValueOnce({
                success: true,
                data: { id: '1', title: 'Test' }
            });

            mockAllAsync.mockResolvedValueOnce({
                success: true,
                data: [{ name: 'sub1' }]
            });

            mockAllAsync.mockResolvedValueOnce({
                success: false,
                error: 'Failed to get variables'
            });

            const result = await getPromptMetadata('1');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Failed to get variables');
        });

        it('should return error when fragments retrieval fails', async () => {
            mockGetAsync.mockResolvedValueOnce({
                success: true,
                data: { id: '1', title: 'Test' }
            });

            mockAllAsync.mockResolvedValueOnce({
                success: true,
                data: [{ name: 'sub1' }]
            });

            mockAllAsync.mockResolvedValueOnce({
                success: true,
                data: [{ name: 'var1', role: 'test', optional_for_user: false }]
            });

            mockAllAsync.mockResolvedValueOnce({
                success: false,
                error: 'Failed to get fragments'
            });

            const result = await getPromptMetadata('1');
            expect(result.success).toBe(false);
            expect(result.error).toBe('Failed to get fragments');
        });
    });

    describe('viewPromptDetails', () => {
        const mockPrompt: PromptMetadata = {
            id: '1',
            title: 'Test Prompt',
            primary_category: 'test',
            subcategories: ['sub1', 'sub2'],
            directory: 'test-dir',
            tags: ['tag1', 'tag2'],
            one_line_description: 'Test description',
            description: 'Full test description',
            variables: [
                { name: 'var1', role: 'test role', optional_for_user: false },
                { name: 'var2', role: 'test role 2', optional_for_user: true }
            ],
            content_hash: 'test-hash',
            fragments: [{ category: 'test', name: 'fragment1', variable: 'var1' }]
        };
        it('should display prompt details correctly', async () => {
            const consoleSpy = jest.spyOn(console, 'log');
            const mockReadEnvVars = readEnvVars as jest.MockedFunction<typeof readEnvVars>;
            mockReadEnvVars.mockResolvedValueOnce({
                success: true,
                data: [{ id: 1, name: 'var1', value: 'test-value', scope: 'global' }]
            });

            await viewPromptDetails(mockPrompt);

            expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('Prompt:'), mockPrompt.title);
        });
    });
});
