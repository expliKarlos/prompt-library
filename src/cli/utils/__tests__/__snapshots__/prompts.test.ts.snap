// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromptsUtils viewPromptDetails should display env variable correctly 1`] = `
"Prompt: Test Prompt

Full test description

Category: Test

Tags: tag1, tag2

Options: ([*] Required  [ ] Optional)
  --var1 [*] 
    test role
      Env: env_var_name (env-value)"
`;

exports[`PromptsUtils viewPromptDetails should display fragment variable correctly 1`] = `
"Prompt: Test Prompt

Full test description

Category: Test

Tags: tag1, tag2

Options: ([*] Required  [ ] Optional)"
`;

exports[`PromptsUtils viewPromptDetails should display prompt details correctly 1`] = `
"Prompt: Test Prompt

Full test description

Category: Test

Tags: tag1, tag2

Options: ([*] Required  [ ] Optional)
  --var1 [*] (Env variable available)
    test role
      Not Set (Required)
  --var2 [ ] 
    test role 2
      Not Set"
`;

exports[`PromptsUtils viewPromptDetails should display regular variable value correctly 1`] = `
"Prompt: Test Prompt

Full test description

Category: Test

Tags: tag1, tag2

Options: ([*] Required  [ ] Optional)"
`;

exports[`PromptsUtils viewPromptDetails should handle env vars fetch failure 1`] = `
"Prompt: Test Prompt

Full test description

Category: Test

Tags: tag1, tag2

Options: ([*] Required  [ ] Optional)
  --var1 [*] 
    test role
      Not Set (Required)
  --var2 [ ] 
    test role 2
      Not Set"
`;
